### DISCUSSION TITLE
Using zod tuples

### DISCUSSION THREAD

**Discussion by augustskare** [edmundhung/conform#287] [General]
Might be a odd use cases, but I need to validate a form where the fields need to be grouped in a given order. To achieve this I'm using zod tuples. Validation works fine and all, but I'm not able to access the field (get errors, generated ids etc.). 
 
```jsx
const schema = z.object({ credentials: z.tuple([z.object({}), z.object({})]) });

const [form, { credentials }] = useForm({
  onValidate({ formData }) {
    return parse(formData, { schema });
  },
});

credentials[0].fieldX // TypeError: Cannot read properties of undefined (reading 'fieldX')
```

I also tried to access it using `useFieldset`, but I don't think that's the right use case:
```jsx
const fields = useFieldset(form.ref, credentials);
const filedset = useFieldset(form.ref, fields[0])
fieldset.fieldX // returns `FieldConfig` but with the incorrect name "credentials.0.fieldX" not "credentials[0].fieldX"
```

**Comment by edmundhung** [comment]
That's indeed an interesting usecase. I did a quick test and this should work for what you need:

```tsx
const schema = z.object({
  kv: z.tuple([z.string(), z.number()]),
});

const [form, { kv }] = useForm({
  defaultValue: {
    // This is necessary to make sure `useFieldList` initialized properly with 2 items
    kv: ['', ''],
  },
  onValidate({ formData }) {
    return parse(formData, { schema });
  },
});
const [key, value] = useFieldList(form.ref, kv);

console.log(key.name) // kv[0]
console.log(value.name) // kv[1]
```

There are 2 minor issues here:
- You must specify a default value as `useFieldList` initialized to empty list by default.
- The inferred type can be improved. It is now considered `string | number` for each item instead of `string` for first item and `number` for the second item. You will need to cast it to `FieldConfig<{ ... }>` manually when dealing with nested object.

I am still trying to understand the implication and the other possible usecases here. But I agree it's not a bad idea to support this with `useFieldset` as well. 

```tsx
// instead of this:
const [key, value] = useFieldList(form.ref, kv);

// Maybe we can do this in the future?
const [key, value] = useFieldset(form.ref, kv);
```

**Reply by augustskare** [reply]
Thanks for pointing me in the right direction! Only change I had to do in my situation (with objects inside tuples) was using `useFieldset` in combination with `useFieldList`. Other then that it worked out nicely 😊 

```jsx
const fieldList = useFieldList(form.ref, kv);
const fields = useFieldset(form.ref, fieldList[index]);
console.log(fields.fieldX) 
```
